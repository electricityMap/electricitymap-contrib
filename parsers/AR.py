#!/usr/bin/env python3


from datetime import datetime, timedelta
from logging import Logger, getLogger
from typing import Dict, List, Optional

import arrow
from pytz import timezone
from requests import Session

from .lib.exceptions import ParserException

# Useful links.
# https://en.wikipedia.org/wiki/Electricity_sector_in_Argentina
# https://en.wikipedia.org/wiki/List_of_power_stations_in_Argentina
# http://globalenergyobservatory.org/countryid/10#
# http://www.industcards.com/st-other-argentina.htm

# API Documentation: https://api.cammesa.com/demanda-svc/swagger-ui.html
CAMMESA_DEMANDA_ENDPOINT = (
    "https://api.cammesa.com/demanda-svc/generacion/ObtieneGeneracioEnergiaPorRegion/"
)

CAMMESA_RENEWABLES_ENDPOINT = "https://cdsrenovables.cammesa.com/exhisto/RenovablesService/GetChartTotalTRDataSource/"

CAMMESA_RENEWABLES_REGIONAL_ENDPOINT = (
    "https://cdsrenovables.cammesa.com/exhisto/RenovablesService/getCapacidadesRegiones"
)

CAMMESA_REGIONS_ENDPOINT = "https://api.cammesa.com/demanda-svc/demanda/RegionesDemanda"

CAMMESA_EXCHANGE_ENDPOINT = (
    "https://api.cammesa.com/demanda-svc/demanda/IntercambioCorredoresGeo/"
)

REGIONS = {
    "AR": "Total del SADI ",
    "AR-BAS": "Buenos Aires",
    "AR-NEA": "NEA",
    "AR-NOA": "NOA",
    "AR-CEN": "Centro",
    "AR-PAT": "Patagonia",
    "AR-LIT": "Litoral",
    "AR-COM": "Comahue",
    "AR-CUY": "Cuyo",
}

SUPPORTED_EXCHANGES = {
    "AR->CL-SEN": "ARG-CHI",
    "AR->PY": "ARG-PAR",
    "AR->UY": "ARG-URU",
    "AR-NEA->BR-S": "ARG-BRA",
    "AR-BAS->AR-COM": "PBA-COM",
    "AR-CEN->AR-COM": "CEN-COM",
    "AR-CEN->AR-NOA": "CEN-NOA",
    "AR-COM->AR-CUY": "CUY-COM",
    "AR-CEN->AR-CUY": "CEN-CUY",
    "AR-LIT->AR-NEA": "LIT-NEA",
    "AR-BAS->AR-LIT": "PBA-LIT",
    "AR-CUY->AR-NOA": "CUY-NOA",
    "AR-CEN->AR-LIT": "LIT-CEN",
    "AR-LIT->AR-NOA": "LIT-NOA",
    "AR-BAS->AR-CEN": "PBA-CEN",
    "AR-COM->AR-PAT": "COM-PAT",
    "AR-NEA->AR-NOA": "NEA-NOA",
}

EXCHANGE_DIRECTIONS = {  # directions are from first region -> second region
    "AR->CL-SEN": 180,
    "AR->PY": 45,
    "AR->UY": 0,
    "AR-NEA->BR-S": 45,
    "AR-BAS->AR-COM": 225,
    "AR-CEN->AR-COM": 270,
    "AR-CEN->AR-NOA": 135,
    "AR-COM->AR-CUY": 135,
    "AR-CEN->AR-CUY": 225,
    "AR-LIT->AR-NEA": 45,
    "AR-BAS->AR-LIT": 90,
    "AR-CUY->AR-NOA": 45,
    "AR-CEN->AR-LIT": 45,
    "AR-LIT->AR-NOA": 135,
    "AR-BAS->AR-CEN": 135,
    "AR-COM->AR-PAT": 270,
    "AR-NEA->AR-NOA": 180,
}


def fetch_production(
    zone_key="AR",
    session: Optional[Session] = None,
    target_datetime: Optional[datetime] = None,
    logger: Logger = getLogger(__name__),
) -> List[dict]:
    """Requests up to date list of production mixes (in MW) of a given region."""

    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    current_session = session or Session()

    non_renewables_production = non_renewables_production_mix(zone_key, current_session)
    renewables_production = renewables_production_mix(zone_key, current_session)

    full_production_list = [
        {
            "datetime": arrow.get(datetime_tz_ar).to("UTC").datetime,
            "zoneKey": zone_key,
            "production": merged_production_mix(
                non_renewables_production[datetime_tz_ar],
                renewables_production[datetime_tz_ar],
            ),
            "capacity": {},
            "storage": {},
            "source": "cammesaweb.cammesa.com",
        }
        for datetime_tz_ar in non_renewables_production
        if datetime_tz_ar in renewables_production
    ]

    return full_production_list


def merged_production_mix(non_renewables_mix: dict, renewables_mix: dict) -> dict:
    """Merges production mix data from different sources. Hydro comes from two
    different sources that are added up."""

    production_mix = {
        "biomass": renewables_mix["biomass"],
        "solar": renewables_mix["solar"],
        "wind": renewables_mix["wind"],
        "hydro": non_renewables_mix["hydro"] + renewables_mix["hydro"],
        "nuclear": non_renewables_mix["nuclear"],
        "unknown": non_renewables_mix["unknown"],
    }

    return production_mix


def renewables_production_mix(zone_key: str, session: Session) -> Dict[str, dict]:
    """Retrieves production mix for renewables using CAMMESA's API"""

    now = datetime.now(tz=timezone("America/Argentina/Buenos_Aires"))
    today = now.strftime("%d-%m-%Y")
    endpoint = CAMMESA_RENEWABLES_REGIONAL_ENDPOINT
    params = {}
    minute = now.minute
    rounded = minute - minute % 5
    time = now.replace(minute=rounded, second=0, microsecond=0) - timedelta(minutes=5)
    region_name = zone_key[3:]

    if zone_key == "AR":
        params = {"desde": today, "hasta": today}
        endpoint = CAMMESA_RENEWABLES_ENDPOINT

    renewables_response = session.get(endpoint, params=params)
    if renewables_response.status_code != 200:
        raise ParserException(
            "AR.py",
            f"Exception when fetching production for {zone_key}: error when calling renewables endpoint: [{renewables_response.status_code}]  {renewables_response.text}",
            zone_key,
        )

    production_list = renewables_response.json()
    sorted_production_list = sorted(production_list, key=lambda d: d["momento"])

    if zone_key != "AR":
        sorted_production_list = list(
            filter(lambda d: d["nemoRegion"] == region_name, production_list)
        )

    renewables_production: Dict[str, dict] = {
        (
            datetime.strptime(
                production_info["momento"], "%Y-%m-%dT%H:%M:%S.%f%z"
            ).isoformat()
            if zone_key == "AR"
            else time.isoformat()
        ): {
            "biomass": production_info["biocombustible"],
            "hydro": production_info["hidraulica"],
            "solar": production_info["fotovoltaica"],
            "wind": production_info["eolica"],
        }
        for production_info in sorted_production_list
    }
    return renewables_production


def non_renewables_production_mix(zone_key: str, session: Session) -> Dict[str, dict]:
    """Retrieves production mix for non renewables using CAMMESA's API"""

    id = get_region_id(zone_key, session)
    params = {"id_region": id}
    api_cammesa_response = session.get(CAMMESA_DEMANDA_ENDPOINT, params=params)
    if api_cammesa_response.status_code != 200:
        raise ParserException(
            "AR.py",
            f"Exception when fetching production for {zone_key}: error when calling non-renewables endpoint: [{api_cammesa_response.status_code}]  {api_cammesa_response.text}",
            zone_key,
        )

    production_list = api_cammesa_response.json()
    sorted_production_list = sorted(production_list, key=lambda d: d["fecha"])

    non_renewables_production: Dict[str, dict] = {
        datetime.strptime(
            production_info["fecha"], "%Y-%m-%dT%H:%M:%S.%f%z"
        ).isoformat(): {
            "hydro": production_info["hidraulico"],
            "nuclear": production_info["nuclear"],
            # As of 2022 thermal energy is mostly natural gas but
            # the data is not split. We put it into unknown for now.
            # More info: see page 21 in https://microfe.cammesa.com/static-content/CammesaWeb/download-manager-files/Sintesis%20Mensual/Informe%20Mensual_2021-12.pdf
            "unknown": production_info["termico"],
        }
        for production_info in sorted_production_list
    }
    return non_renewables_production


def get_region_id(zone_key, session: Session) -> int:
    """Fetches the region id for the zone that is required to get the production of that zone."""
    regions_response = session.get(CAMMESA_REGIONS_ENDPOINT)

    if regions_response.status_code != 200:
        raise ParserException(
            "AR.py",
            f"Exception when fetching regions for AR: error when calling regions endpoint: [{regions_response.status_code}]  {regions_response.text}",
            zone_key,
        )

    regions = regions_response.json()
    region_to_id = {region["nombre"]: region["id"] for region in regions}

    region_name = REGIONS[zone_key]
    return region_to_id[region_name]


def fetch_exchange(
    zone_key1: str,
    zone_key2: str,
    session: Optional[Session] = None,
    target_datetime: Optional[datetime] = None,
    logger: Logger = getLogger(__name__),
) -> dict:
    """Requests the last known power exchange (in MW) between two zones."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    sorted_zone_keys = sorted([zone_key1, zone_key2])
    sorted_codes = "->".join(sorted_zone_keys)
    flow: Optional[float] = None
    returned_datetime: datetime
    exchange = {}

    if sorted_codes in SUPPORTED_EXCHANGES:
        current_session = session or Session()

        api_cammesa_response = current_session.get(CAMMESA_EXCHANGE_ENDPOINT)
        if api_cammesa_response.status_code != 200:
            raise ParserException(
                "AR.py",
                f"Exception when fetching exchange for {sorted_codes}: error when calling exchange endpoint: [{api_cammesa_response.status_code}]  {api_cammesa_response.text}",
                sorted_zone_keys,
            )

        exchange_name = SUPPORTED_EXCHANGES[sorted_codes]
        exchange_list = api_cammesa_response.json()
        for exchange in exchange_list["features"]:
            properties = exchange["properties"]
            if properties["nombre"] == exchange_name:
                angle_config = EXCHANGE_DIRECTIONS[sorted_codes]
                given_angle = int(properties["url"][6:])
                flow = int(properties["text"])
                if angle_config != given_angle:
                    flow = -flow
                returned_datetime = datetime.fromisoformat(
                    properties["fecha"][:-2] + ":" + properties["fecha"][-2:]
                )
                if flow is None:
                    raise ParserException(
                        "AR.py",
                        f"Failed fetching exchange for {sorted_zone_keys}",
                        sorted_codes,
                    )
                exchange = {
                    "sortedZoneKeys": sorted_codes,
                    "datetime": returned_datetime,
                    "netFlow": flow,
                    "source": "cammesaweb.cammesa.com",
                }
    else:
        raise NotImplementedError("This exchange is not currently implemented")

    return exchange


def fetch_price(
    zone_key: str = "AR",
    session: Optional[Session] = None,
    target_datetime: Optional[datetime] = None,
    logger: Logger = getLogger(__name__),
) -> dict:
    """Requests the last known power price of a given country."""

    raise NotImplementedError("Fetching the price is not currently implemented")


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
    print("fetch_price() ->")
    print(fetch_price())
    print("fetch_exchange(AR, PY) ->")
    print(fetch_exchange("AR", "PY"))
    print("fetch_exchange(AR, UY) ->")
    print(fetch_exchange("AR", "UY"))
    print("fetch_exchange(AR, CL-SEN) ->")
    print(fetch_exchange("AR", "CL-SEN"))
