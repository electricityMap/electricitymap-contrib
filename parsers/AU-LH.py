from datetime import datetime, timezone
from logging import Logger, getLogger
from typing import Any, Dict, List, Optional, Union
import time

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    PriceList,
    ProductionBreakdownList,
    TotalConsumptionList,
    TotalProductionList,
)
from electricitymap.contrib.lib.models.events import ProductionMix, StorageMix
from electricitymap.contrib.lib.types import ZoneKey
from parsers.lib.exceptions import ParserException

from requests_html import HTMLSession


PARSER_NAME = "AU-LH.py"
DATA_URL = "http://photonscada.com/data/perspective/client/LHI"

RENDER_TIMEOUT = 10 # seconds

def validate_and_clean_data(
    zone_key: ZoneKey,
    values: List[str]
) -> List[float]:
    if len(values) != 8:
        raise ParserException(PARSER_NAME, "Unexpected number of values parsed", zone_key)
    
    even_index_values = values[::2]
    if not all(val == "L" for val in even_index_values):
        raise ParserException(PARSER_NAME, "Unexpected parsed values (although normally discarded), website structure changed", zone_key)

    odd_index_values = values[1::2]
    cleaned_values = []

    for val in odd_index_values:
        # Observed units are always kW, this could be a wrong assumption
        unit_suffix = " kW"
        if not val.endswith(unit_suffix):
            raise ParserException(PARSER_NAME, f"Unexpected parsed values, check website structure or electricity units {odd_index_values}", zone_key)
        
        # strip suffix, convert to numerical value, convert units to MW
        parsed_val = float(val.rstrip(unit_suffix))/1000
        num_digits = 10
        cleaned_values.append(round(parsed_val, num_digits))

    return cleaned_values


def fetch_data(
    zone_key: ZoneKey,
    session: HTMLSession
) -> Dict[str, float]:
    res = session.get(DATA_URL)

    if not res.status_code == 200:
        raise ParserException(
            PARSER_NAME,
            f"Exception when fetching production error code: {res.status_code}",
            zone_key,
        )
    
    # render the website using request_html (chromium), need a sleep value to wait for render
    res.html.render(timeout=RENDER_TIMEOUT + 1, sleep=RENDER_TIMEOUT)

    # values in the DOM are poorly structured
    # thankfully the values we want are returned in tspans
    tspan_list = res.html.find("tspan")
    parsed_values = [tspan.text for tspan in tspan_list]

    cleaned_values = validate_and_clean_data(zone_key, parsed_values)

    return {
        "consumption": cleaned_values[0],
        "solar": cleaned_values[1],
        "battery": cleaned_values[2],
        "diesel": cleaned_values[3]
    }
    

def fetch_production(
    zone_key: ZoneKey,
    session: HTMLSession = None,
    target_datetime: Optional[datetime] = None,
    logger: Logger = getLogger(__name__),
) -> Union[List[dict], dict]:
    if target_datetime is not None:
        raise ParserException(PARSER_NAME, "This parser is not yet able to parse past dates", zone_key)
    
    is_new_session = session is None
    if is_new_session:
        session = HTMLSession()

    data = fetch_data(zone_key, session)

    production_list = ProductionBreakdownList(logger=logger)
    production_list.append(
        zoneKey=zone_key,
        datetime=datetime.now(timezone.utc),
        production=ProductionMix(
            gas=data["diesel"],
            solar=data["solar"],
        ),
        # storage=StorageMix(hydro=item["hydroStorage"] * -1),
        source=DATA_URL,
    )

    if is_new_session:
        session.close()

    return production_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Maps backend, but handy for testing."""

    print(fetch_production(ZoneKey("AU-LH")))
