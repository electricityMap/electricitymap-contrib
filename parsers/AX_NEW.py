from datetime import datetime
from logging import Logger, getLogger
from typing import Optional
from requests import Response, Session
from bs4 import BeautifulSoup
from re import findall

IFRAME_URL = "https://grafik.kraftnat.ax/grafer/tot_inm_24h_15.php"


def fetch_data(session: Session, logger: Logger):
    """Fetch data from the iFrame."""

    response: Response = session.get(IFRAME_URL)
    response.raise_for_status()

    soup = BeautifulSoup(response.text, "html.parser").find_all("script")
    time_series = findall(r"data: \[(.*?)\]", str(soup))[0]
    raw_data = findall(r"data:\[(.*?)\]", str(soup))
    data_list = []
    for time, sweden, alink, fossil, gustavs, wind, consumption in zip(
        time_series.split(","),
        raw_data[0].split(","),
        raw_data[1].split(","),
        raw_data[2].split(","),
        raw_data[3].split(","),
        raw_data[4].split(","),
        raw_data[5].split(","),
    ):
        data_list.append(
            {
                "time": str(time.replace('"', "")),
                "sweden": float(sweden),
                "alink": float(alink),
                "fossil": float(fossil),
                "gustavs": float(gustavs),
                "wind": float(wind),
                "consumption": float(consumption),
            }
        )
    return data_list


def fetch_production(
    zone_key: str = "AX",
    session: Session = Session(),
    target_datetime: Optional[datetime] = None,
    logger: Logger = getLogger(__name__),
) -> dict:
    """Fetch production from the iFrame."""

    if target_datetime is not None:
        raise NotImplementedError(
            "The datasource currently implemented is only real time"
        )

    data = fetch_data(session, logger)

    return data
