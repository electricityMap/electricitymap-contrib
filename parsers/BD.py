#!/usr/bin/env python3

from datetime import datetime, timedelta
from logging import Logger, getLogger
from typing import Any
from zoneinfo import ZoneInfo

import arrow
from bs4 import BeautifulSoup, Tag
from requests import Response, Session

from parsers.lib.config import refetch_frequency
from parsers.lib.exceptions import ParserException

# Power Grid Company of Bangladesh: erp.pgcb.gov.bd
# Has table (also historical) of production, consumption and exchange.
# This table gets updated batch-wise every few hours, so most of the time, the delay will be >2h.

tz = ZoneInfo("Asia/Dhaka")
latest_url = "https://erp.pgcb.gov.bd/web/generations/view_generations"
url_by_date = (
    "https://erp.pgcb.gov.bd/web/generations/view_generations?search="  # DD-MM-YYYY
)

TABLE_HEADERS = [
    "Date",
    "Time",
    "Generation(MW)",
    "Demand(MW)",
    "Loadshed",
    "Gas",
    "Liquid Fuel",
    "Coal",
    "Hydro",
    "Solar",
    "Wind",
    "Bheramara HVDC",
    "Tripura",
    "Remarks",
]


def table_entry_to_float(entry: str):
    """
    Helper function to handle empty cells in table.
    """
    if entry == "":
        return None
    try:
        return float(entry)
    except ValueError:
        raise ParserException(
            parser="BD.py",
            message=(f'Failed to parse entry: "{entry}" to float in table.'),
        )


def parse_table_body(table_body: Tag):
    """
    Parse the table body returned by the URL.
    Returns a list of rows represented by dicts.
    """

    rows = table_body.find_all("tr")
    row_data = []

    for row in rows:
        row_items = row.find_all("td")
        row_items = [item.text.strip() for item in row_items]

        row_dict = dict()

        # date and time in [0]; [1] are DD-MM-YYYY; HH:mm[:ss]
        parsed_day = arrow.get(row_items[0], "DD-MM-YYYY", tzinfo=tz).datetime
        try:
            parsed_time = arrow.get(row_items[1], "HH:mm:ss", tzinfo=tz).time()
        except arrow.parser.ParserMatchError:
            # very old data points are sometimes in HH:mm format
            parsed_time = arrow.get(row_items[1], "HH:mm", tzinfo=tz).time()

        row_dict["time"] = datetime.combine(parsed_day, parsed_time, tzinfo=tz)

        # [2] is total generation in MW
        row_dict["total_generation"] = table_entry_to_float(row_items[2])
        # [3] is total demand in MW
        row_dict["total_demand"] = table_entry_to_float(row_items[3])
        # [4] is loadshed, not interesting for us
        row_dict["loadshed"] = table_entry_to_float(row_items[4])
        # [5] is generation from gas
        row_dict["gas"] = table_entry_to_float(row_items[5])
        # [6] is generation from liquid fuel (oil)
        row_dict["oil"] = table_entry_to_float(row_items[6])
        # [7] is generation from coal
        row_dict["coal"] = table_entry_to_float(row_items[7])
        # [8] is generation from hydro
        row_dict["hydro"] = table_entry_to_float(row_items[8])
        # [9] is generation from solar
        row_dict["solar"] = table_entry_to_float(row_items[9])
        # [10] is generation from wind
        row_dict["wind"] = table_entry_to_float(row_items[10])
        # [11], [12] is import from Bheramara PP (IN-EA) and Tripura (IN-NE)
        row_dict["bd_import_bheramara"] = table_entry_to_float(row_items[11])
        row_dict["bd_import_tripura"] = table_entry_to_float(row_items[12])
        row_dict["remarks"] = row_items[13]

        row_data.append(row_dict)

    return row_data


def verify_table(table_header: Tag):
    """
    Validate the table headers, that it looks like expected.
    Don't parse if the table has changed.
    """
    header_items = table_header.find_all("th")
    header_items = [item.text.strip() for item in header_items]

    if header_items != TABLE_HEADERS:
        raise ParserException(
            parser="BD.py",
            message=(
                f"Table headers mismatch with expected ones."
                f"Expected: {TABLE_HEADERS}"
                f"Parsed: {header_items}"
            ),
        )


def query(
    session: Session, target_datetime: datetime | None, logger: Logger
) -> list[dict[str, Any]]:
    """
    Query the table and read it into list.
    """
    # build URL to call
    if target_datetime is None:
        target_url = latest_url
    else:
        # Convert timezone of target_datetime, and build URL from day
        target_datetime_bd = arrow.get(target_datetime).to(tz)
        target_dt_str = target_datetime_bd.format("DD-MM-YYYY")
        target_url = url_by_date + target_dt_str

    target_response: Response = session.get(target_url)

    if not target_response.ok:
        raise ParserException(
            parser="BD.py",
            message=f"Data request did not succeed: {target_response.status_code}",
        )

    response_soup = BeautifulSoup(target_response.text)

    # Find the table, there is only one, and verify its headers.
    table = response_soup.find("table")
    if table is None:
        raise ParserException(
            parser="BD.py",
            message="Could not find table in returned HTML.",
        )

    table_head = table.find("thead")
    if table_head is None:
        raise ParserException(
            parser="BD.py",
            message=("Could not find table header in returned HTML."),
        )
    verify_table(table_head)

    # Table valid as we expect, parse the rows.
    table_body = table.find("tbody")
    if table_body is None:
        raise ParserException(
            parser="BD.py",
            message=("Could not find table body in returned HTML."),
        )
    row_data = parse_table_body(table_body)

    return row_data


@refetch_frequency(timedelta(days=1))
def fetch_production(
    zone_key: str = "BD",
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> dict[str, Any] | list[dict[str, Any]]:
    row_data = query(session, target_datetime, logger)

    production_data_list = []
    for row in row_data:
        # Create data with empty production
        data = {
            "zoneKey": zone_key,
            "datetime": row["time"],
            "production": {},
            "source": "erp.pgcb.gov.bd",
        }

        # And add sources if they are present in the table
        known_sources_sum_mw = 0.0
        for source_type in ["coal", "gas", "hydro", "oil", "solar", "wind"]:
            if row[source_type] is not None:
                # also accumulate the sources to infer 'unknown'
                known_sources_sum_mw += row[source_type]
                data["production"][source_type] = row[source_type]

        # infer 'unknown'
        if row["total_generation"] is not None:
            unknown_source_mw = row["total_generation"] - known_sources_sum_mw
            if unknown_source_mw >= 0:
                data["production"]["unknown"] = unknown_source_mw
            else:
                logger.warn(
                    f"Sum of production sources exceeds total generation by {-unknown_source_mw}MW."
                    f"There is probably something wrong..."
                )

        production_data_list.append(data)

    if not len(production_data_list):
        raise ParserException(
            parser="BD.py",
            message="No valid consumption data for requested day found.",
        )
    return production_data_list


@refetch_frequency(timedelta(days=1))
def fetch_consumption(
    zone_key: str = "BD",
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> dict[str, Any] | list[dict[str, Any]]:
    row_data = query(session, target_datetime, logger)

    result_list = []

    for row in row_data:
        # get the demand from the table
        consumption = row["total_demand"]
        if consumption is None:
            continue  # no data in table

        result_list.append(
            {
                "datetime": row["time"],
                "consumption": consumption,
                "zoneKey": zone_key,
                "source": "erp.pgcb.gov.bd",
            }
        )

    if not len(result_list):
        raise ParserException(
            parser="BD.py",
            message="No valid consumption data for requested day found.",
        )

    return result_list


@refetch_frequency(timedelta(days=1))
def fetch_exchange(
    zone_key1: str,
    zone_key2: str,
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    # Query table, contains import from india.
    row_data = query(session, target_datetime, logger)

    result_list = []
    sortedZoneKeys = "->".join(sorted([zone_key1, zone_key2]))

    for row in row_data:
        # BD -> IN_xx
        if zone_key2 == "IN-NE":
            # Export to India NorthEast via Tripura
            bd_import = row["bd_import_tripura"]
        elif zone_key2 == "IN-EA":
            # Export to India East via Bheramara
            bd_import = row["bd_import_bheramara"]
        else:
            raise ParserException(
                parser="BD.py",
                message=f"Exchange pair {sortedZoneKeys} is not implemented.",
            )

        if bd_import is None:
            continue  # no data in table
        bd_export = -1.0 * bd_import

        result_list.append(
            {
                "sortedZoneKeys": sortedZoneKeys,
                "datetime": row["time"],
                "netFlow": bd_export,
                "source": "erp.pgcb.gov.bd",
            }
        )

    return result_list


if __name__ == "__main__":
    print("fetch_production() ->")
    print(fetch_production())
    print("fetch_consumption() ->")
    print(fetch_consumption())
    print("fetch_exchange('BD', 'IN-NE') ->")
    print(fetch_exchange("BD", "IN-NE"))
    print("fetch_exchange('BD', 'IN-EA') ->")
    print(fetch_exchange("BD", "IN-EA"))
