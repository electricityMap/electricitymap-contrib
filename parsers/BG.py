from datetime import datetime, timezone
from logging import Logger, getLogger

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionBreakdown, ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.exceptions import ParserException

PARSER = "BG.py"

SOURCE = "eso.bg"
SOURCE_API_URL = "http://www.eso.bg/api/rabota_na_EEC_json.php"

PRODUCTION_TYPES_BG_TO_EN = {
    "АЕЦ": "nuclear",
    "Кондензационни ТЕЦ": "coal",
    "Топлофикационни ТЕЦ": "gas",
    "Заводски ТЕЦ": "gas",
    "ВЕЦ": "hydro",
    "Малки ВЕЦ": "hydro",
    "ВяЕЦ": "wind",
    "ФЕЦ": "solar",
    "Био ТЕЦ": "biomass",
    "Био ЕЦ": "biomass",
    "Товар РБ": "consumption",
}


def _fetch_all_production_events(
    zone_key: ZoneKey,
    session: Session,
    target_datetime: datetime | None,
    logger: Logger,
) -> list[ProductionBreakdown]:
    """Fetches all production events from the backend source API."""
    if target_datetime is not None:
        raise ParserException(
            PARSER, "This parser is not yet able to parse historical data", zone_key
        )

    time = datetime.now(timezone.utc)
    response = session.get(SOURCE_API_URL)
    if not response.status_code == 200:
        raise ParserException(
            PARSER,
            f"Exception when fetching production error code: {response.status_code}: {response.text}",
            zone_key,
        )

    response_payload = response.json()
    logger.debug(f"Raw generation breakdown: {response_payload}")

    production = {}
    for header, value in response_payload:
        for production_type_bg in PRODUCTION_TYPES_BG_TO_EN:
            if header.startswith(production_type_bg):
                production_type_en = PRODUCTION_TYPES_BG_TO_EN[production_type_bg]
                production[production_type_en] = (
                    production.get(production_type_en, 0.0) + value
                )

    event = ProductionBreakdown.create(
        logger=logger,
        zoneKey=zone_key,
        datetime=time,
        source=SOURCE,
        production=ProductionMix(
            biomass=production.get("biomass"),
            coal=production.get("coal"),
            gas=production.get("gas"),
            geothermal=production.get("geothermal"),
            hydro=production.get("hydro"),
            nuclear=production.get("nuclear"),
            oil=production.get("oil"),
            solar=production.get("solar"),
            unknown=production.get("unknown"),
            wind=production.get("wind"),
        ),
        storage=None,
    )
    return [event]


def fetch_production(
    zone_key: ZoneKey = ZoneKey("BG"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    session = session or Session()
    all_production_events = _fetch_all_production_events(
        zone_key=zone_key,
        session=session,
        target_datetime=target_datetime,
        logger=logger,
    )
    production_list = ProductionBreakdownList(logger)
    for event in all_production_events:
        production_list.append(
            zoneKey=event.zoneKey,
            datetime=event.datetime,
            source=event.source,
            production=event.production,
            storage=event.storage,
            sourceType=event.sourceType,
        )
    return production_list.to_list()
