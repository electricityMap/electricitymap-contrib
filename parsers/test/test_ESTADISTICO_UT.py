from importlib import resources
from requests import Session
from requests_mock import GET, Adapter, POST
from snapshottest import TestCase

from electricitymap.contrib.lib.types import ZoneKey
from parsers.ESTADISTICO_UT import fetch_production, DAILY_OPERATION_URL


class TestESTADISTICO_UT(TestCase):
    def setUp(self) -> None:
        self.session = Session()
        self.adapter = Adapter()
        self.session.mount("https://", self.adapter)
        self.session.mount("http://", self.adapter)
        self.adapter.register_uri(
            GET,
            DAILY_OPERATION_URL,
            text=resources.files("parsers.test.mocks.ESTADISTICO_UT")
                .joinpath("production.html")
                .read_text(),
        )
        self.adapter.register_uri(
            POST,
            DAILY_OPERATION_URL,
            text=resources.files("parsers.test.mocks.ESTADISTICO_UT")
                .joinpath("data.html")
                .read_text(),
        )

    def test_fetch_production(self):
        production = fetch_production(
            zone_key=ZoneKey("SV"),
            session=self.session,
        )

        self.assertMatchSnapshot(
            [
                {
                    "datetime": element["datetime"].isoformat(),
                    "production": element["production"],
                    "storage": element["storage"],
                    "source": element["source"],
                    "zoneKey": element["zoneKey"],
                }
                for element in production
            ]
        )
