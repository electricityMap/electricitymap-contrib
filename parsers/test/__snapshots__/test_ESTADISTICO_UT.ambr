# serializer version: 1
# name: test_fetch_production
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 1, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='America/El_Salvador')),
      'production': dict({
        'biomass': 145.81,
        'geothermal': 162.33,
        'hydro': 160.81,
        'unknown': 58.28,
        'wind': 18.08,
      }),
      'source': 'ut.com.sv',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'SV',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 1, 1, 0, tzinfo=zoneinfo.ZoneInfo(key='America/El_Salvador')),
      'production': dict({
        'biomass': 147.53,
        'geothermal': 165.11,
        'hydro': 129.55,
        'unknown': 59.52,
        'wind': 15.79,
      }),
      'source': 'ut.com.sv',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'SV',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 1, 2, 0, tzinfo=zoneinfo.ZoneInfo(key='America/El_Salvador')),
      'production': dict({
        'biomass': 148.04,
        'geothermal': 166.12,
        'hydro': 97.36,
        'unknown': 56.4,
        'wind': 19.01,
      }),
      'source': 'ut.com.sv',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'SV',
    }),
  ])
# ---
